typescript/sdk/src/providers/MultiProvider.ts

```diff
@@ -8,104 +8,53 @@ import {
   providers,
 } from 'ethers';
 
-import { Address, exclude, pick } from '@hyperlane-xyz/utils';
+import { Address, pick } from '@hyperlane-xyz/utils';
 
 import { chainMetadata as defaultChainMetadata } from '../consts/chainMetadata';
 import { CoreChainName, TestChains } from '../consts/chains';
-import {
-  ChainMetadata,
-  getDomainId,
-  isValidChainMetadata,
-} from '../metadata/chainMetadataTypes';
+import { ChainMetadataManager } from '../metadata/ChainMetadataManager';
+import { ChainMetadata } from '../metadata/chainMetadataTypes';
 import { ChainMap, ChainName } from '../types';
 
-import { RetryJsonRpcProvider, RetryProviderOptions } from './RetryProvider';
+import {
+  DEFAULT_RETRY_OPTIONS,
+  ProviderBuilderFn,
+  defaultProviderBuilder,
+} from './providerBuilders';
 
 type Provider = providers.Provider;
 
-const DEFAULT_RETRY_OPTIONS: RetryProviderOptions = {
-  maxRequests: 3,
-  baseRetryMs: 250,
-};
-
-export function defaultProviderBuilder(
-  rpcUrls: ChainMetadata['rpcUrls'],
-  network: providers.Networkish,
-  retryOverride?: RetryProviderOptions,
-): Provider {
-  const createProvider = (r: ChainMetadata['rpcUrls'][number]) => {
-    const retry = r.retry || retryOverride;
-    return retry
-      ? new RetryJsonRpcProvider(retry, r.http, network)
-      : new providers.StaticJsonRpcProvider(r.http, network);
-  };
-  if (rpcUrls.length > 1) {
-    return new providers.FallbackProvider(rpcUrls.map(createProvider), 1);
-  } else if (rpcUrls.length === 1) {
-    return createProvider(rpcUrls[0]);
-  } else {
-    throw new Error('No RPC URLs provided');
-  }
-}
-
-export type ProviderBuilderFn = typeof defaultProviderBuilder;
-
-interface MultiProviderOptions {
+export interface MultiProviderOptions {
   loggerName?: string;
-  providerBuilder?: ProviderBuilderFn;
+  providerBuilder?: ProviderBuilderFn<Provider>;
 }
 
-export class MultiProvider {
-  public readonly metadata: ChainMap<ChainMetadata> = {};
-  protected readonly providers: ChainMap<Provider> = {};
-  protected readonly providerBuilder: ProviderBuilderFn;
-  protected signers: ChainMap<Signer> = {};
-  protected useSharedSigner = false; // A single signer to be used for all chains
-  protected readonly logger: Debugger;
+/**
+ * A utility class to create and manage providers and signers for multiple chains
+ * @typeParam MetaExt - Extra metadata fields for chains (such as contract addresses)
+ */
+export class MultiProvider<MetaExt = {}> extends ChainMetadataManager<MetaExt> {
```

Comment:
did we add the default type for backwards compatibility?

Reply:
That and convenience